src/Security/Authentication/Core/src/AuthAppBuilderExtensions.cs

```diff
@@ -24,7 +24,13 @@ public static IApplicationBuilder UseAuthentication(this IApplicationBuilder app
             throw new ArgumentNullException(nameof(app));
         }
 
-        app.Properties[AuthenticationMiddlewareSetKey] = true;
-        return app.UseMiddleware<AuthenticationMiddleware>();
+        // Don't add more than one instance of the middleware
+        if (!app.Properties.ContainsKey(AuthenticationMiddlewareSetKey))
+        {
+            app.Properties[AuthenticationMiddlewareSetKey] = true;
+            return app.UseMiddleware<AuthenticationMiddleware>();
+        }
```

Comment:
This check shouldn't be there. The user needs to be able to re-run auth if they add this to the pipeline again.

Reply:
Why would you re-run auth?

Reply:
That's exactly why you'd want to re-run it later in the pipeline to force setting the User.

Reply:
I guess your point is that there's never a situation where running UseAuthentication can change based on where it is in the pipeline. Is that right?

Reply:
```C#
app.UseAuthentication();

app.Use((context, next) => 
{    
    context.User = new ClaimsPrincipal();
});
app.UseAuthentication(); // Should this noop?
```

This is breaking and we should avoid breaking this. While this might seem non-sensical, its impossible to understand which components might be calling UseAuthentication to force it to re-run.

Reply:
The trade-off is perf. If the middleware is added automatically at the start, and then added manually again later then it always runs twice and does duplicate work. How can someone that wants to manually add the middleware prevent the automatic one?

Reply:
> How can someone that wants to manually add the middleware prevent the automatic one?

They can avoid it by not interacting with `WebApplicationBuilder.Authentication` and instead calling their methods directly on `WebApplicationBuilder`.

Reply:
Can we add a flag to let them explicitly turn it off?  `WebApplicationBuider.Authenticaiton.AutoAddMiddleware = false`?

Reply:
The performance of auth already sucks, and this doesn't make it better enough to warrant the breaking change.

Reply:
> Can we add a flag to let them explicitly turn it off? WebApplicationBuider.Authenticaiton.AutoAddMiddleware = false?

That feels a little strange to me. The `WebApplicationBuilder.Authentication` pattern is an opt-in to the more simplified approach. It feels weird to have an opt-out to that when the more obvious choice of not using the new `Authentication` property at all exists.

Reply:
> How can someone that wants to manually add the middleware prevent the automatic one?

Our check in the `WebApplicationBuilder` prevents this from happening. We'll set a flag in `UseAuthentication` that is read and avoids re-registering automatically in the WAB.

Reply:
> That feels a little strange to me. The WebApplicationBuilder.Authentication pattern is an opt-in to the more simplified approach. It feels weird to have an opt-out to that when the more obvious choice of not using the new Authentication property at all exists.

It feels strange to me that there'd be scenarios we're consciously aware of where we'd say the workaround is to **not** use the new property.  Can we simply make it so that the `WebApplicationBuilder` does not add the middlewares if the app pipeline has already added the authentication middleware? Why is any more than that needed?

Reply:
> Can we simply make it so that the WebApplicationBuilder does not add the middlewares if the app pipeline has already added the authentication middleware? Why is any more than that needed?

It already does that. To clarify, my point was that we didn't need to do anything additional here.